ledger/participant-integration-api/src/main/scala/platform/store/backend/h2/H2StorageBackend.scala

```diff
@@ -197,7 +197,32 @@ private[backend] object H2StorageBackend
       connectionInitHook: Option[Connection => Unit],
   )(implicit loggingContext: LoggingContext): DataSource = {
     val h2DataSource = new org.h2.jdbcx.JdbcDataSource()
-    h2DataSource.setUrl(jdbcUrl)
+
+    // H2 (org.h2.jdbcx.JdbcDataSource) does not support setting the user/password within the jdbcUrl, so remove
+    // those properties from the url if present and set them separately. This helps run canton tests which set
+    // user/password. Note that Postgres and Oracle support user/password in the URLs, so we don't bother exposing
+    // user/password configs separately from the url just for h2 which is anyway not supported for production.
+    def setKeyValueAndRemoveFromUrl(url: String, key: String, setter: String => Unit): String = {
+      val separator = ";"
+      url.toLowerCase.indexOf(separator + key + "=") match {
+        case -1 => url
+        case i =>
+          val valueBegin = i + 1 + key.length + 1 // separator, key, "="
+          val (value, shortenedUrl) = url.indexOf(separator, i + 1) match {
+            case -1 => (url.substring(valueBegin), url.take(i))
+            case next =>
+              (url.substring(valueBegin, next), url.take(i) + url.takeRight(url.length - next))
+          }
```

Comment:
Wouldn't it be easier to match for regex `s";${key}=([^;]*)"`?

Reply:
I actually looked into using regex as the first solution. I believe that would work as well although it seemed as I would have to make two separate regex calls - one to extract the value and another to `Regex.replaceFirst`.

Reply:
The regex I mentioned should give you two results: the first the substring that needs to be removed (can be passed to `String.replace`), the second the extracted value (can be passed to `setter`). I didn't actually try it though, so not sure if it would be actually better.

Reply:
I think it is a great idea to use regex here as well, if we can!

Reply:
Yes we can: Here is the regex version that seems to work when I tried manually:

```
  def setKeyValueAndRemoveFromUrl(url: String, key: String, setter: String => Unit): String = {
    val regex = new Regex(s".*(;${key}=([^;])).*")
    url match {
      case regex(keyAndValue, value) =>
        setter(value)
        url.replace(keyAndValue, "")
      case _ => url
    }
  }
```